JPA 로 전환 - Member

soosue opened the pull request:
안녕하세요 :)
이번에 PR 은 member 패키지에 대해 리팩토링을 진행해봤습니다.

말이 길어져서, 잘 안 읽히실 것 같아 줄로 문단을 나눠봤는데, 읽는데 도움이 되셨으면 좋겠습니다. :)

앞으로의 방향성(?) 같은 피드백도 주시면 도움이 될 것 같습니다.
피드백 잘 부탁드립니다. 😄

closes #7 
참고 #4

soosue committed:
refactor: Member 테이블 쿼리 새로운 칼럼으로 변경

기존 Member 테이블 칼럼명이 가독성이 좋지 않아
보기 편한 칼럼명으로 변경

soosue committed:
refactor: Member 회원가입 로직 수정 (JPA)

1. Member 엔티티 생성
2. Member 저장 로직 JPA 로 변경 (saveMember 메서드)
3. Oracle 이므로 sequence 사용

soosue committed:
refactor: 회원가입 email 확인 url 변경

기존: /member/emailCheck.do
변경 후: /v1/member/check

soosue committed:
refactor: 회원가입 email 확인 로직 정리

기존 ModelAndView 와 response 를 직접 응답해주었으나
불필요한 관계로 간단하게 정리

soosue committed:
refactor: 회원가입 email 확인, JPA 사용하도록 변경

soosue committed:
refactor: jsp 에서 email 확인하는 로직 정리

1. 불필요한 코드 제거
2. 중복 코드 추출

soosue committed:
chore: jsp 코드 정리

soosue committed:
refactor: 로그인 기능 JPA 적용

soosue committed:
refactor: 로그아웃 기능 수정

soosue committed:
chore: member 패키지 메서드명, 불필요 메서드 정리

soosue committed:
chore: Member 테이블 칼럼 변경에 맞게 DDL, DML 수정

soosue committed:
refactor: kakao 로그인 기능 수정

soosue committed:
refactor: memberService transactional 어노테이션 추가

soosue committed:
chore: 패키지 명 변경

변경 전: com.java.guesthouse.api.login
변경 후: com.java.guesthouse.api.oauth

soosue committed:
refactor: kakao 로그아웃 accessToken 만료 시키는 API 호출 추가

soosue committed:
chore: member 관련 jsp 정리

soosue committed:
chore: jsp 패키지 명 오타 수정

변경 전: tempalte
변경 후: template

soosue committed:
refactor: MemberService 인터페이스 제거

soosue committed:
chore: member register page 요청하는 url 주소 변경, 가독성 위한 줄바꿈

soosue committed:
refactor: Member 객체 테스트 추가

비밀번호가 동일한지 확인하는 테스트

soosue committed:
chore: profile 과 카카오 OAuth 로그인 가짜 rest-api-key 추가

해당 설정값이 존재하지 않아서 서버 시작시 에러 발생
실제 애플리케이션에서는 application-key.yml 파일 존재하며
실제 카카오 OAuth 로그인 rest api 키 값으로 설정되어있음

soosue commented:
---
DTO 는 Java 17 의 record 를 이용하여 변경하였습니다.

---
service, dao interface 제거하였고,

---
적절한 변수 명과 메서드 명을 지으려고 해봤습니다.

다만 메서드 명을 지을 때,
`반환 값이 있는 메서드 명`은 `명사로 지어도 된다`라는 규칙을 추가할까 생각하는데 의견이 있으실까요?

`MemberController.registerMemberPage()` 의 기능은 회원가입 페이지로 이동하는 것인데,
동사로 이름 지으려면 `goToRegisterMemberPage()` 와 같은 이름이 되어야할 것 같은데 괜찮은지 모르겠습니다.
프론트엔드가 추가 된다면 필요 없는 메서드가 될 것 같긴 합니다.

그리고, record 를 이용하면 getter 메서드가 필드 뒤에 소괄호 () 가 붙은 형식으로 생성되더라고요. 그래서

`메서드 명은 반환 값이 있을 때는 반환 값에 적절한 명사형을 허용한다.` vs `record 의 경우, 명사형을 허용한다.` 에 대해 고민을 했었습니다. 저는 둘 다 괜찮다고 생각했습니다.

---
그 외에 service 단에서는 dto(XXXrequest) 를 이용하고, repository 에 접근할 때는 엔티티를 이용하게 바꿨는데요.
Member 쪽은 도메인 로직이 복잡한 것이 없어서 그런지 도메인 객체와 엔티티 객체를 분리해야 하는 경우는 발생하지 않았던 것 같습니다.
그 예시를 볼 수 없어 아쉬웠습니다.

---
jsp 는 나름 정리하고 있는데, 주석이라던가 조금 남아있습니다. 신경 안 써주셔도 괜찮을 것 같습니다!

---
이번에 많이 고민했던 부분은 "테이블 칼럼 명이 변경되는 경우 어떻게 해야 하는가?" 였습니다.

기존 Member 테이블의 칼럼 명이 맘에 들지 않아 변경을 하려고 봤더니, 실제 운영 상황이라면 어떻게 해야 할까 생각했더니 재밌었습니다.

![image](https://user-images.githubusercontent.com/54942017/195538986-9f73ab0f-2573-4160-8a3f-8b25ffcdac3e.png)

위와 같이 진행하면 되겠다고 생각을 했었습니다. 어떻게 생각하시나요?

다른 고민 되는 점이 있었다면

1. 운영에서 칼럼 명이 변경되는 경우가 있긴 할까?
2. 운영 데이터베이스에는 데이터가 엄청 많을텐데 칼럼 하나를 추가하는데 걸리는 시간이 얼마나 걸릴지 모른다는 점.
3. 기존 칼럼 값으로 새로운 칼럼을 update 할 때, DB 락이 걸리지 않을까 하는 생각이었습니다.

3번의 경우는, 10,000개의 데이터가 있다면 1000개씩 분할 업데이트를 하면 해결할 수 있을 것 같았습니다만
다른 고민들은 해결이 잘 되지 않았네요. 😭

---
아 그리고 카카오 API 를 이용하는 부분도 궁금한 점이 있습니다.

저는 api.oauth 패키지에 따로 빼놨는데, 이 패키지의 KakaoOAuth 를 controller 와 service 에서 호출을 하고 있습니다.
service 에만 의존 되게 작성해야 할 것 같은 생각이 근거 없이 마구 드는데... 잘 모르겠습니다. (뭔가 bad smell 같습니다. ㅠㅠ)
bad smell 이 맞을까요?

---
마지막으로 리팩토링을 하고 있기는 한데,
이게 현재 넥스트스텝 챌린지 코스가 생각했던 방향이 아닌 것 같다는 생각이 들어서 조바심이 납니다.
지금 단계에서는 tdd 관련된 리팩토링을 하고 있어야 하는데 그렇지 않은 것 같고, 또 스스로 많이 더딘 것 같습니다.

성현님의 생각은 어떠신가요??

boorownie commented on src/main/java/com/java/guesthouse/member/ui/MemberController.java:
```diff
@@ -2,28 +2,36 @@
-    public ModelAndView memberRegister(HttpServletRequest request, HttpServletResponse response) {
-        System.out.println("Member Register");
+    @RequestMapping(value = "/member/register.page", method = RequestMethod.GET)
+    public ModelAndView registerMemberPage(HttpServletRequest request) {
```
goToRegisterMemberPage로 하지 않고 registerMemberPage로 표현해도 동사를 이용한다는 기존 컨벤션에 부합되는거 같아요!
이 상황과 별개로 메서드 네이밍 컨벤션으로 리턴이 있는 메서드의 경우 명사로 지어도된다는 규칙은 나쁘지 않다고 생각합니다!
조건으로 메서드 내부에서 변환이 있으면 안된다는 추가 규칙 정도만 있어도 좋겠네요 :)

soosue replied:
메서드 내부에 변환이 있으면 안된다는 규칙은 불변객체 느낌을 말하는 것 맞으실까요?

boorownie replied:
꼭 불변 객체가 아니더라도 해당 메서드 내에서는 객체의 상태 변화가 일어나지 않는다는 것이 보장 되면 좋겠다는 의미였습니다.

boorownie commented on src/main/java/com/java/guesthouse/member/domain/Member.java:
```diff
@@ -0,0 +1,64 @@
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "MEMBER_SEQ_GEN")
+    private Long id;
+
+    private String name;
```
제 경험으로는 규모에 따라 다를 순 있지만 운영에서 memberName -> name으로 변경되더라도 여러가지 위험요소를 가지면서까지 테이블의 컬럼명까지 변경하진 않을 것 같습니다ㅎㅎ 그 이유는 10000개 혹은 더 많은 데이터가 있을텐데 1000개씩 분할해가면서 까지 이름을 변경하는 장점을 공감하지 못해서입니다ㅎㅎ

데이터베이스를 마이그레이션 하는것은 상당히 고려해야할 게 많더라구요

그럼에도 불구하고 만약 컬럼명 변경을 한다면 저라면 이렇게 할 것 같습니다.

새로운 테이블을 별도로 생성하고 애플리케이션에서 기존 테이블과 신규 테이블에 양쪽에 동시에 insert하도록 합니다. 그런다음 기존 데이터를 신규 테이블에 나눠서 붓습니다. 붓는 동안 이슈가 있을 수 있으니 여러번 시도하고 싱크를 맞추는 작업을 합니다. 그런 다음읽기 작업을 새로운 테이블로 하도록 합니다. 장애가 생기면 언제든지 기존 테이블을 바라보게 준비를 해둡니다. 그런다음 안정화가 되면 기존 테이블을 제거합니다.

물론 이 방법도 헛 점이 있고 더 나은 방법이 있겠지만 안전하고 덜 복잡하게 옮기는 방법이 아닐까 생각이 듭니다ㅎㅎ

soosue replied:
역시 리스크가 있군요. 😠
데이터를 붓는걸 잘 해야할 것 같습니다. 감사합니다 :)

boorownie commented on src/main/java/com/java/guesthouse/member/ui/MemberController.java:
```diff
@@ -32,87 +40,63 @@ public ModelAndView memberRegister(HttpServletRequest request, HttpServletRespon
+        HttpSession session = request.getSession();
+        String accessToken = (String)session.getAttribute("accessToken");
+        if (accessToken != null) {
+            kakaoOAuth.logout(accessToken);
```
kakaoOAuth에 대한 의존이 꺼려진다면 loginService와 같은 객체에 위임해서 수행하는건 어떠신가요??

soosue replied:
새로운 loginService 객체를 만들어서 역할을 나누면서 의존을 제거해보라는 말씀이신거죠? 적용해보겠습니다.

boorownie commented on src/main/resources/mapper/AdminMapper.xml:
```diff
@@ -1,77 +1,123 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
jpa를 사용하게 된다면 mapper 파일들은 제거가 가능한 상태인가요?

soosue replied:
네 지금은 위에서 말씀드린 것처럼 member 패키지에 대해서만 리팩토링을 했고, member 패키지의 MemberMapper.xml 은 제거가 된 상태입니다.

다만 AdminMapper 에 Member 와 관련된 sql이 남아있는걸 제대로 확인하지 못했네요.
Member 와 관련된 sql들은 제거해보겠습니다. :)

boorownie commented on src/main/java/com/java/guesthouse/search/service/SearchService.java:
```diff
@@ -5,7 +5,8 @@
 import com.java.guesthouse.host.service.dto.HostDto;
 
 public interface SearchService {
```
인터페이스를 남겨두신건가요?

soosue replied:
모두 제거하도록 하겠습니다!

boorownie commented on src/main/java/com/java/guesthouse/search/service/SearchServiceImp.java:
```diff
@@ -32,12 +32,12 @@ public SearchServiceImp(SearchDao searchDao) {
     @Override
-    public ModelAndView search(String checkIn, String checkOut, String local, String people, String searchHouseName, String pageNumber, Integer memberCode, String sort) {
-        HomeAspect.logger.info(HomeAspect.logMsg + "local: " + local + ", checkIn: " + checkIn + ", checkOut: " + checkOut + " ,people: " + people + ", searchHouseName: " + searchHouseName + ", pageNumber: " + pageNumber + ", memberCode: " + memberCode + ", sort: " + sort);
+    public ModelAndView search(String checkIn, String checkOut, String local, String people, String searchHouseName,
```
Service레이어에서 ModelAndView 객체를 리턴하게되면
Service에서 Presentation 객체에 대한 의존이 생기게됩니다.
dto를 리턴하는 방법은 어떠실까요?

soosue replied:
네 저도 동의합니다. 하나씩 바꿔가겠습니다 :)

boorownie commented on src/main/java/com/java/guesthouse/member/service/MemberService.java:
```diff
@@ -1,16 +1,95 @@
+        return member.getId();
+    }
+
+    @Transactional(readOnly = true)
```
class 차원에서 `@Transactional(readOnly = true)`을 설정해두고
필요한 메서드에만 `@Transactional`을 설정하면 어떨까요?

soosue replied:
반대로 설정하지 않는 이유가 있을까요??

boorownie replied:
꼭 코멘트 드린대로 할 필요는 없지만
코멘트 드린것 처럼 할 경우 실수로 @Transactional을 붙이지 않은 경우 테스트로 잡아낼 수 있는 반면
반대로 할 경우 테스트로도 의도하지 않은 설정을 확인하기 어려울 수 있습니다.

boorownie commented on src/main/java/com/java/guesthouse/member/service/MemberService.java:
```diff
@@ -1,16 +1,95 @@
+    }
+
+    @Transactional(readOnly = true)
+    public void login(LoginRequest loginRequest, HttpServletRequest request)
```
business layer 객체에서 presentation layer의 객체를 의존하고 있네요!
HttpServletRequest 객체를 파라미터로 받기 보다는 member객체를 응답해주어 controller에서 session에 저장해주면 좋을 것 같아요.

boorownie commented on src/main/java/com/java/guesthouse/member/service/MemberService.java:
```diff
@@ -1,16 +1,95 @@
+        session.setAttribute("memberLevel", member.getMemberLevel());
+        session.setAttribute("email", member.getEmail());
+        session.setAttribute("memberCode", member.getId());
+    }
```
이 부분도 controller의 역할로 옮기면 좋겠어요!

boorownie commented on src/main/java/com/java/guesthouse/member/service/MemberService.java:
```diff
@@ -1,16 +1,95 @@
+    }
+
+    @Transactional
+    public void kakaoLogin(KakaoLoginRequest kakaoLoginRequest, HttpServletRequest request) {
```
business layer 객체에서 presentation layer의 객체를 의존하고 있네요!

boorownie commented on src/main/java/com/java/guesthouse/api/oauth/GetTokenApi.java:
```diff
@@ -0,0 +1,37 @@
+                .body(BodyInserters
+                        .fromFormData("grant_type", "authorization_code")
+                        .with("client_id", restApiKey)
+                        .with("redirect_uri", "http://localhost:8080/v1/member/kakaologin")
```
redirect_uri를 비롯하여 grant_type도 상수로 분리해도 좋을 것 같아요!

soosue replied:
[0b8e751](https://github.com/soosue/my_guesthouse/pull/8/commits/0b8e751184e40236e5dfd2b53235668ff8d7b21f) 이렇게 변경해봤습니다.

`"grant_type", "client_id", "redirect_uri"` 키에 해당하는 부분은 공통적인 부분도 없기 때문에
직접 입력하는게 좋다고 생각했는데 어떻게 생각하시나요?

boorownie replied:
아 코멘트의 의미도 값만 상수화 하는게 어떨지 제안드린겁니다.
만약 키도 상수화를 한다면 HttpStatus와 같이 상수 클래스로 분리해서 사용할 것 같습니다만 
저라면 안 그럴 것 같습니다ㅎㅎㅎ

boorownie commented on src/main/java/com/java/guesthouse/api/oauth/GetUserInfoApi.java:
```diff
@@ -0,0 +1,32 @@
+@Component
+public class GetUserInfoApi {
+    private static final HttpMethod METHOD = HttpMethod.POST;
+    private static final String HOST = "https://kapi.kakao.com";
+    private static final String URI = "/v2/user/me";
```
GetUserInfoApi 외 GetTokenApi 등 url을 상수로 가지고 있는데요~
카카오와 관련된 url을 관리하는 상수 객체를 두는건 어떨까요?

soosue replied:
[3422655](https://github.com/soosue/my_guesthouse/pull/8/commits/3422655576442e770c58754215871e761fa58885) 좋은 의견 같아서 별도로 분리해봤습니다.

uri 외에도 공통적으로 사용할 수 있는 부분도 같이 분리해봤습니다. 현재는 복잡하지 않아서 다 같이 한 곳에 몰아도 괜찮겠다고 생각했습니다.

boorownie reviewed (changes requested):
jpa 전환 잘 해주셨습니다!
사실상 jpa로의 전환 보다는 이로인한 패키지 리팩터링에 대한 작업이 더 많아 보이네요!ㅎㅎ

프로젝트는 점진적으로 발전하고 있는것 같아서 긍정적으로 생각하고 있습니다!
윤수님이 생각하시는 챌린지코스가 생각했던 방향이 어떤 방향인지는 모르겠지만
저는 점진적으로 조금씩 발전해나가는 모습을 봐서는 아주 좋은 방향으로 가고 있다고 생각합니다.
가이드로 나간 챌린지가 있어도 이를 위해서 기반 작업이 수행되어있지 않다면 먼 길을 돌아가야할 수 있는데
그 작업을 먼저 수행하고 계신거라 생각이 듭니다.

우선 눈에 보이는 큰 줄기의 코멘트 위주로 남겼습니다.
코멘트가 달린 부분만 개선하기 보다는
코멘트가 의미하는바를 이해하신 후 관련된 부분도 함께 개선해보면 더 좋을 것 같아요!

soosue committed:
refactor: uri 규칙에 따라 엔티티를 나타내는 단어를 복수형으로 변경

soosue committed:
refactor: 로그아웃 기능 LoginService 에게 위임

MemberController 가 kakaoOAuth 와 의존성이 있었고, 이를 더 적절한 객체와 의존을 맺게 하고 싶음

적절한 역할을 하는 LoginService 를 만들어 역할을 위임하고 의존성을 제거함

soosue committed:
chore: 불필요한 코드 제거

soosue committed:
refactor: AdminMapper.xml 에서 member 관련 sql JPA 로 전환

soosue committed:
refactor: AdminMapper.xml 에서 member sql 완전히 제거

- AdminMapper.xml 에서 member 쿼리 제거 완료
- Members 조회 페이징, data jpa Pageable 이용
- Members 조회 rest API 조회로 변경

soosue committed:
refactor: 인터페이스 제거

soosue committed:
refactor: business layer 객체에서 presentation layer의 객체 의존 제거

- service 에서 HttpServletRequest 인자로 받지 않게 변경
- 역할이 더 맞는 LoginService 로 메서드 이동

soosue committed:
refactor: API parameter 값을 상수로 변경

soosue committed:
refactor: KakaoOAuth API 관련 값을 상수로 관리

boorownie reviewed (approved):
변경된 내용 확인 했고 추가 질문 주신 부분은 코멘트에 답변 남겼습니다!
전반적으로 리뷰 반영이 코멘트 드린 방향대로 잘 된것 같습니다 :) 

참고로 리뷰 요청 시 질문은 한 pr 메시지에 남기는 것 보다
해당 코드의 코멘트로 남겨주시면 제가 확인하는데 조금 더 수월할 것 같습니다 🙇‍♂️

고생 많으셨습니다~
